BrainPortal/lib/boutiques_input_copier.rb

```diff
@@ -0,0 +1,308 @@
+      new_input["default-value"] = config["default-value"] || false
+      hide                       = config["hide"]          || false
+      new_input.delete("command-line-flag")
+
+      copied_info[parent_inputid] = { input_idx: index+1, new_input: new_input, hide: hide}
+    end
+
+    # Get the position in the groups where the new inputs will be inserted
+    copied_info.each do |parent_inputid, info|
+      descriptor.groups.each do |group|
+        group["members"].each_with_index do |member, index|
+          next if member != parent_inputid
+          copied_info[parent_inputid][:member_info] = {idx: index + 1, id: group["id"]}
+        end
+      end
+    end
+
+    copied_info
+  end
+
+  # Adjust the descriptor for the input with the mention of
+  # the fake parent directory information
+  #
+  # For each input in BoutiquesInputCopier section create a new entry in the descriptor
+  # that will be a flag to choose if the input will be copied if the checkbox is selected
+  def descriptor_with_special_input(descriptor) #:nodoc:
+    parent_input_ids = descriptor.custom_module_info('BoutiquesInputCopier')
+
+    inputs      = descriptor["inputs"]
+
+    # In parent_input_ids, select all non hidden inputs
+    parent_input_ids = parent_input_ids.select { |_, config| !config["hide"] }
+    copied_info      = fill_copied_info(parent_input_ids, descriptor, inputs)
+
+    # Add the new inputs to the descriptor and to the group if needed
+    extracted_idx = copied_info.values.map { |info| info[:input_idx] }.sort.reverse
+
+    extracted_idx.each do |index|
+      # In copied_info, find the one with input_idx == index
+
+      copied_info.each do |parent_inputid, info|
+        next if info[:hide]
+        next if info[:input_idx] != index
+
+        # Insert the new input in the descriptor
+        new_input = info[:new_input]
+        inputs.insert(index, new_input)
+
+        # Insert the new input in the group
+        next if !info[:member_info]
+        group        = descriptor.group_by_id(info[:member_info][:id])
+        members      = group["members"]
+        new_input_id = new_input["id"]
+        next if members.include?(new_input_id)
+        members.insert(info[:member_info][:idx], new_input_id)
+      end
+    end
+
+    descriptor
+  end
+
+  def descriptor_for_before_form #:nodoc:
+    descriptor_with_special_input(super.dup)
+  end
+
+  def descriptor_for_form #:nodoc:
+    descriptor_with_special_input(super.dup)
+  end
+
+  def descriptor_for_after_form #:nodoc:
+    descriptor_with_special_input(super.dup)
+  end
+
+  ############################################
+  # Bourreau (Cluster) Side Modifications
+  ############################################
+
+  # For input in +BoutiquesInputCopier+ section,
+  def setup #:nodoc:
+    original_userfile_ids = {}
+
+    basename = Revision_info.basename
+    commit   = Revision_info.short_commit
+    self.addlog("Creating a copy of the input files in BoutiquesInputCopier.")
```

Comment:
I don't like the order of things happening in this method. super is called too low, AFTER the messages that says "Creating a copy", which means that if there is a crash in any other code it looks like this module is responsible. Not onlhy that, it means the "copy" message will be separated with the other addlog messages by all the other messages produced by the other modules.e.g.

```
Creation copy.... (message from this module)
(messages from other modules)
(messages from other modules)
More messages from this module (if any)